### ISSUE TITLE
Rule proposal: `prefer-private-class-fields`

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#979]
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

Prefer private fields over the convention of underscore prefixed names.

This fearure requires Node.js 12, so we cannot add it to the recommended preset until April 2021 (date of Node.js 10 obsoletion).

## Fail

```js
class Foo {
	static _PRIVATE_STATIC_FIELD

	_privateField

	_privateMethod() {
		return 'hello world'
	}
}
```

## Pass

```js
class Foo {
	static #PRIVATE_STATIC_FIELD

	#privateField

	#privateMethod() {
		return 'hello world'
	}
}
```

**Comment by sindresorhus** [comment]
This is now accepted.

---

But it's blocked by:

> Does ESLint support this feature yet?

**Comment by fregante** [comment]
[No longer blocked](https://eslint.org/demo#eyJ0ZXh0IjoiY2xhc3MgQSB7XG5cdCNuYW1lID0gMVxufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWVsc2UtaWYiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWltcG9ydC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1sb3NzLW9mLXByZWNpc2lvbiI6Miwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tbm9ub2N0YWwtZGVjaW1hbC1lc2NhcGUiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXByb3RvdHlwZS1idWlsdGlucyI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zZXR0ZXItcmV0dXJuIjoyLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVuc2FmZS1vcHRpb25hbC1jaGFpbmluZyI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtYmFja3JlZmVyZW5jZSI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7fX19)

**Comment by fisker** [comment]
I've already tried to implement this the other day, my experience on using private fields, I really need [proposal-destructuring-private](https://github.com/tc39/proposal-destructuring-private)

```js
const {a, b, c, d} = this
const e = this.#e
```

awkward!

**Comment by sindresorhus** [comment]
@fisker I would argue having to use a `_` prefix to pseudo hide properties/methods is also awkward. I don't think lack of destructuring is awkward enough to not use private fields. It's just a minor inconvenience. The `proposal-destructuring-private` thing will probably not land in Node.js until Node.js 20 at the earliest, which means packages [will not be able to use it until 2025](https://github.com/nodejs/Release).

**Comment by bbugh** [comment]
One thing to consider is that `#private` fields and proxies do not work together. References:

1. https://github.com/tc39/proposal-class-fields/issues/106
2. https://github.com/evanw/esbuild/issues/1969

**Comment by jimmywarting** [comment]
> in regards to destructuring-private

i have also had to do:
```js
const a = this.#a
const b = this.#b
const c = this.#c
```
And I agree that it feels a bit awkward and inconvenient. but not something that's bothering me that much. 
i did found that it could be just simpler to just have one single private property bag where i then store all my things
```js
const {a, b, c} = this.#bag || { }
```

i rather prefer `#` over soft `private` keyword like in typescript that don't change anything.
esbuild can mangle private class fields down to just 1-2 letter words resulting in better compression and things stay private as it should be doing. and it's easier to recognize that some random piece of code is making use of private class fields in some middle of the code.

**Comment by jimmywarting** [comment]
```js
{
  'no-restricted-syntax': [
    'error',
    {
      selector: ':matches(PropertyDefinition, MethodDefinition)[accessibility="private"]',
      message: 'Use #private instead',
    },
  ],
}
```
[playground](https://typescript-eslint.io/play/#ts=4.6.4&sourceType=module&showAST=es&code=PQKgBApgzgNglgOwC5gQewLQCdpK3AYyQgBMMoBPZAQwA8AuMAbQCgx2wByCLLNLTgBo2HAN4iOHKBBgQi-Rp3oBbakgIALaAAoACnwAOPJBQAiEAGaI4SOGgSCwAWQhINaEuasIbdhAEomagICaCg4ACM4eBMAXgAiA3wANzUIeIBdIQlJMGUw6gBzCEUAVWkwAGIkuFTiMEQoYmoSbMkAX2EMsBBgFhYCGGooKDAAMTQ0MHFJJLQDAEZGagQKAG4cgwBXCPgCMDmDACZl1Y3ZvmIiUgPDAGZT9c2UtNv5gBZH845qwwBWL79ST5NweBbafzTdqbHZ7PKudwkI4QqHPNBXYgkeGgkh3FGiaEXWqvEGI974wk-UkeP4UoEcYooYKhEb8cGQ0RgHBILZYBBgBZgNaU9jbXaEMCMsDMsL8ZEcrmuXn8wXCtEYm5SmWsrB4hXc5UCoUi27E+pakKyrDk-VKvlGtWSKWVbVQfi0208+2q6HQoA&rules=N4XyA&tsConfig=N4XyA)
